### PR TITLE
[core][node] Support ImageSource

### PR DESCRIPTION
**Pull request by asheemmamoowala**
Add support for ImageSource to mbgl-core. The current source passes existing render tests and can be verified with the macos app/mbgl-glfw.

## Remaining Tasks
- [x] Improve render transform
- [x] Fix image flip when crossing date line 🙃 
- [x] ImageSource API improvements
- ~iOS/Android/Qt Bindings ? ⛷~  (Working on a separate PR)
- [x] Offline support
- [x] rebase 💥 

//cc @jfirebaugh @ivovandongen @kkaefer

### CODE REVIEW

File: cmake/core-files.cmake [hunk 1]
```diff
@@ -163,79 +163,87 @@ set(MBGL_CORE_FILES
...
     src/mbgl/renderer/render_tile.hpp
-    src/mbgl/renderer/symbol_bucket.cpp
-    src/mbgl/renderer/symbol_bucket.hpp
     src/mbgl/renderer/tile_parameters.hpp
     src/mbgl/renderer/tile_pyramid.cpp
     src/mbgl/renderer/tile_pyramid.hpp
     src/mbgl/renderer/transitioning_property.hpp
     src/mbgl/renderer/update_parameters.hpp
 
+    # renderer/buckets
```
Comment (jfirebaugh) [inline cmake/core-files.cmake]: 👍 on subdirectories for buckets and layer. Can you submit that as an independent PR so that it's easily reviewable independently? Will also help minimize potential merge conflicts as these files are getting touched a lot right now.

File: include/mbgl/style/conversion/coordinate.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+#include <mbgl/style/conversion.hpp>
+#include <mbgl/util/geo.hpp>
+
+namespace mbgl {
+namespace style {
+namespace conversion {
+            
+template<>
+struct Converter<std::unique_ptr<LatLng>> {
```
Comment (jfirebaugh) [inline include/mbgl/style/conversion/coordinate.hpp]: No need for `unique_ptr` here; `LatLng` is a value class.

File: include/mbgl/util/geo.hpp [hunk 3]
```diff
@@ -47,7 +47,7 @@ class LatLng {
             wrap();
         }
     }
-
+
```
Comment (jfirebaugh) [inline include/mbgl/util/geo.hpp]: Nit: whitespace change.

File: platform/default/mbgl/storage/offline_download.cpp [hunk 4]
```diff
@@ -107,6 +107,7 @@ OfflineRegionStatus OfflineDownload::getStatus() const {
 
         case SourceType::Video:
         case SourceType::Annotations:
+        case SourceType::Image:
             break;
```
Comment (jfirebaugh) [inline platform/default/mbgl/storage/offline_download.cpp]: We'll need to add offline support for image sources -- it should download and store the image in the offline database.

File: platform/default/mbgl/storage/offline_download.cpp [hunk 5]
```diff
@@ -176,7 +177,8 @@ void OfflineDownload::activateDownload() {
 
             case SourceType::Video:
             case SourceType::Annotations:
-                break;
+            case SourceType::Image:
+                    break;
```
Comment (jfirebaugh) [inline platform/default/mbgl/storage/offline_download.cpp]: Whitespace.

File: src/mbgl/renderer/painters/painter_raster.cpp [hunk 6]
```diff
@@ -42,7 +42,8 @@ static std::array<float, 3> spinWeights(float spin) {
 void Painter::renderRaster(PaintParameters& parameters,
                            RasterBucket& bucket,
                            const RenderRasterLayer& layer,
-                           const RenderTile& tile) {
+                           const mat4& matrix,
+                           bool useBucketBuffers = false) {
```
Comment (asheemmamoowala) [inline src/mbgl/renderer/painters/painter_raster.cpp]: @kkaefer - The painter methods are layer specific as opposed to source specific. Adding a separate painter function would also impact the RasterBucket which calls this method. 
Changing the method signature here woks for all RasterBuckets and RasterLayers, as opposed to creating a cascade of additional changes. 
It also allows us to reuse this painter to render raster layers with other non-tiled sources.

File: include/mbgl/style/sources/image_source.hpp [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace mbgl {
+class LatLng;
+class AsyncRequest;
+
+namespace style {
+
+class ImageSource : public Source {
+public:
+    ImageSource(std::string id, const std::vector<LatLng>);
```
Comment (incanus) [inline include/mbgl/style/sources/image_source.hpp]: How come this isn't a `LatLngBounds`?

Comment (1ec5) [inline include/mbgl/style/sources/image_source.hpp]: Apparently an image source can be associated with any quadrilateral, not necessarily a rectangle.

mbgl tends to use `std::array` to represent fixed-length collections in style values (cf. padding and offsets). As @incanus points out, a struct with named fields would be another option that would avoid corner-swapping problems, although maybe we could make it so that the input order doesn’t matter.

**Comment by incanus** [comment]
Quick & dirty iOS bindings for a near-term use case I have, per chat with @asheemmamoowala: 

https://github.com/mapbox/mapbox-gl-native/compare/1350-image-source...jrm-ios-image-source-hack

File: include/mbgl/style/sources/image_source.hpp (hunk 7 above)
Comment (asheemmamoowala) [inline include/mbgl/style/sources/image_source.hpp]: @incanus - thats an excellent point. Ive tried to wrap around that in the SDKs in the [draft API PR](https://github.com/mapbox/mapbox-gl-native/pull/9110). If the Quad struct makes sense, I think it might make more sense to move it into native.

Comment (asheemmamoowala) [inline include/mbgl/style/sources/image_source.hpp]: @incanus - thats an excellent point. I've tried to improve that in the SDKs in the [draft API PR](https://github.com/mapbox/mapbox-gl-native/pull/9110). If the Quad struct makes sense there, I might even move it into native to avoid confusion

Comment (1ec5) [inline include/mbgl/style/sources/image_source.hpp]: Note that there would have to be separate types for mbgl and for the iOS/macOS SDK, because the iOS/macOS SDKs [can’t expose any C++ to the public](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/ios/DEVELOPING.md#making-any-symbol-public).

File: platform/default/mbgl/storage/offline_download.cpp (hunk 4 above)
Comment (asheemmamoowala) [inline platform/default/mbgl/storage/offline_download.cpp]: Offline support added in [`3506bb5`](https://github.com/mapbox/mapbox-gl-native/pull/8968/commits/3506bb5cf752412095353336e44c0eee806fe6bb)

File: src/mbgl/style/sources/image_source_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+class ImageSource::Impl : public Source::Impl {
+public:
+    Impl(std::string id, std::vector<LatLng> coords);
+    Impl(const Impl& rhs, std::vector<LatLng> coords);
+    Impl(const Impl& rhs, UnassociatedImage image);
+
+    ~Impl() final;
+
+    void setImage(UnassociatedImage&& );
```
Comment (jfirebaugh) [inline src/mbgl/style/sources/image_source_impl.hpp]: Unused.

File: include/mbgl/style/sources/image_source.hpp [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    ~ImageSource() override;
+
+    const std::string& getURL() const;
+    void setURL(const std::string& url) ;
+
+    void setImage(mbgl::UnassociatedImage&&);
```
Comment (jfirebaugh) [inline include/mbgl/style/sources/image_source.hpp]: Nit: no `mbgl::` prefix here.

File: src/mbgl/style/sources/image_source.cpp [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    url = std::move(url_);
+    // Signal that the source description needs a reload
+    if (loaded || req) {
+        loaded = false;
+        req.reset();
+        observer->onSourceDescriptionChanged(*this);
+    }
+}
+
+void ImageSource::setImage(mbgl::UnassociatedImage&& image_) {
```
Comment (jfirebaugh) [inline src/mbgl/style/sources/image_source.cpp]: No `mbgl::` prefix.

File: src/mbgl/style/sources/image_source_impl.hpp [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    const mbgl::UnassociatedImage& getImage() const;
```
Comment (jfirebaugh) [inline src/mbgl/style/sources/image_source_impl.hpp]: No `mbgl::` prefix.

File: include/mbgl/style/sources/image_source.hpp [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace style {
+
+class ImageSource : public Source {
+public:
+    ImageSource(std::string id, const std::vector<LatLng>);
+    ~ImageSource() override;
+
+    const std::string& getURL() const;
+    void setURL(const std::string& url) ;
```
Comment (jfirebaugh) [inline include/mbgl/style/sources/image_source.hpp]: Nit: no space before `;`.

File: src/mbgl/renderer/buckets/raster_bucket.cpp [hunk 13]
```diff
@@ -1,26 +1,41 @@
...
 }
 
+RasterBucket::RasterBucket(RasterBucket&& other) : image(std::move(other.image)) {
+    uploaded = false;
+}
+
 void RasterBucket::upload(gl::Context& context) {
-    texture = context.createTexture(std::move(image));
+    texture = context.createTexture(image);
+    if (vertices.vertexSize() > 0) {
```
Comment (jfirebaugh) [inline src/mbgl/renderer/buckets/raster_bucket.cpp]: `if (!vertices.empty())`

File: src/mbgl/renderer/buckets/raster_bucket.hpp [hunk 14]
```diff
@@ -1,22 +1,38 @@
...
+    void render(Painter& painter,
+                PaintParameters& parameters,
+                const RenderLayer& layer,
+                const mat4& matrix);
     bool hasData() const override;
 
     UnassociatedImage image;
     optional<gl::Texture> texture;
+
+    gl::VertexVector<RasterLayoutVertex> vertices;
```
Comment (jfirebaugh) [inline src/mbgl/renderer/buckets/raster_bucket.hpp]: Can you add a comment here explaining that these are used only when the layer has an image source?

File: src/mbgl/renderer/sources/render_image_source.cpp [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+    GeometryCoordinates geomCoords;
+    for (auto latLng : coords) {
+        auto tc = TileCoordinate::fromLatLng(0, latLng);
+        auto gc = TileCoordinate::toGeometryCoordinate(tileIds[0], tc.p);
+        geomCoords.push_back(gc);
+    }
+    setupBucket(geomCoords);
+}
+
+void RenderImageSource::setupBucket(GeometryCoordinates& geomCoords) {
```
Comment (jfirebaugh) [inline src/mbgl/renderer/sources/render_image_source.cpp]: This is only called from `updateTiles` -- inline it there.

File: src/mbgl/renderer/sources/render_image_source.hpp [hunk 16]
```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <mbgl/renderer/buckets/raster_bucket.hpp>
```
Comment (jfirebaugh) [inline src/mbgl/renderer/sources/render_image_source.hpp]: Forward declare `RasterBucket` (and define `~RenderImageSource` in the .cpp file) instead.

File: src/mbgl/renderer/sources/render_image_source.hpp [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+#include <mbgl/renderer/render_source.hpp>
+#include <mbgl/renderer/render_tile.hpp>
+#include <mbgl/style/sources/image_source_impl.hpp>
+#include <mbgl/tile/geometry_tile_data.hpp>
```
Comment (jfirebaugh) [inline src/mbgl/renderer/sources/render_image_source.hpp]: Unused.

File: src/mbgl/renderer/sources/render_image_source.hpp [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+#include <mbgl/util/image.hpp>
```
Comment (jfirebaugh) [inline src/mbgl/renderer/sources/render_image_source.hpp]: Unused.

File: src/mbgl/renderer/sources/render_image_source.hpp [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+#include <mbgl/util/optional.hpp>
```
Comment (jfirebaugh) [inline src/mbgl/renderer/sources/render_image_source.hpp]: Unused.

File: src/mbgl/renderer/layers/render_raster_layer.cpp [hunk 20]
```diff
@@ -31,4 +36,35 @@ bool RenderRasterLayer::hasTransition() const {
     return unevaluated.hasTransition();
 }
 
+void RenderRasterLayer::uploadBuckets(gl::Context& context, RenderSource* source) {
+    if (renderTiles.size() > 0) {
```
Comment (asheemmamoowala) [inline src/mbgl/renderer/layers/render_raster_layer.cpp]: There is a higher likelihood for the source to be a tile-based `RasterSource`. This order reduces the need for `dynamic_cast` for the majority of cases when `renderTiles` are available.

Comment (jfirebaugh) [inline src/mbgl/renderer/layers/render_raster_layer.cpp]: Did you profile it?

Comment (asheemmamoowala) [inline src/mbgl/renderer/layers/render_raster_layer.cpp]: Reversing the conditional here would incur an unnecessary `dynamic_cast` for the most frequent use cases for a raster layer which will tend to be raster tile source. 

👍 - for base class usage.

**Comment by 1ec5** [comment]
This PR along with #9110 fixes #1350.
